fix(service-worker): `Cache-Control: no-cache` on assets breaks service worker

alan-agius4 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
At the moment `cacheAge` can be undefined and when it is so, it causes `cannot read `length` of `undefined` when having `Cache-Control` set to `no-cache` due the mapping method in `needToRevalidate`


Issue Number: N/A


## What is the new behavior?
No error

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Closes #25442

alan-agius4 commented on packages/service-worker/worker/src/assets.ts:
```diff
@@ -193,9 +193,10 @@ export abstract class AssetGroup {
 
       // Find the max-age directive, if one exists.
-      const cacheAge = cacheDirectives.filter(v => v[0] === 'max-age').map(v => v[1])[0];
+      const maxAge = cacheDirectives.find(v => v[0] === 'max-age');
```
Note: At the moment current it will fail when having 
```js
const cacheAge = [["no-cache"]].filter(v => v[0] === 'max-age').map(v => v[1])[0]
cacheAge.length
```

JoostK replied:
Could this be added as test 👍

alan-agius4 replied:
I am trying to fix how to test it to be honest :) any help?

JoostK replied:
@alan-agius4 I am not familiar with the service worker code at all, sorry. At first glance, looks like we can pull out quite a bit to a separate pure function to easily test:

```typescript
export function parseCacheControlIntoMaxAge(cacheControl: string) {
      const cacheDirectives =
          cacheControl
              // Directives are comma-separated within the Cache-Control header value.
              .split(',')
              // Make sure each directive doesn't have extraneous whitespace.
              .map(v => v.trim())
              // Some directives have values (like maxage and s-maxage)
              .map(v => v.split('='));

      // Lowercase all the directive names.
      cacheDirectives.forEach(v => v[0] = v[0].toLowerCase());

      // Find the max-age directive, if one exists.
      const maxAge = cacheDirectives.find(v => v[0] === 'max-age');
      return maxAge ? maxAge[1] : undefined;
}
```

alan-agius4 replied:
I managed to test it. Thanks anyways. 

Also for me personally, I don't like to export a function just to be able to unit test it

alxhub replied:
I'm okay with exporting it just to test it. Exporting it from a private module within the package is different than exporting it as public API.

alan-agius4 replied:
Yep you are absolutely right.

alan-agius4 commented:
@alxhub, can you have a look at this please?

alxhub reviewed (approved):
TYVM

alan-agius4 commented:
Note: I just changed the commit msg and referenced an issue.

Can we please add the relative `labels` to get this merged?

Thanks.

gkalpak commented on packages/service-worker/worker/src/assets.ts:
```diff
@@ -193,9 +193,10 @@ export abstract class AssetGroup {
 
       // Find the max-age directive, if one exists.
-      const cacheAge = cacheDirectives.filter(v => v[0] === 'max-age').map(v => v[1])[0];
+      const maxAge = cacheDirectives.find(v => v[0] === 'max-age');
```
There is a `maxAge` variable declared few lines below. Consider giving them different names to avoid confusion.

alan-agius4 replied:
@gkalpak, I updated it.

alan-agius4 committed:
fix(service-worker): `Cache-Control: no-cache` on assets breaks service worker

At the moment `cacheAge` can we undefined when having `Cache-Control` set to `no-cache` due the mapping method in `needToRevalidate`

Closes #25442

IgorMinar commented:
caretaker note: no g3 impact

varadero commented on packages/service-worker/worker/src/assets.ts:
```diff
@@ -193,9 +193,10 @@ export abstract class AssetGroup {
       // Find the max-age directive, if one exists.
-      const cacheAge = cacheDirectives.filter(v => v[0] === 'max-age').map(v => v[1])[0];
+      const maxAgeDirective = cacheDirectives.find(v => v[0] === 'max-age');
+      const cacheAge = maxAgeDirective ? maxAgeDirective[1] : undefined;
```
```maxAgeDirective[1]``` ?
